lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnReaderFactory.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.parquet.PrimitiveField;
+import io.trino.spi.TrinoException;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.DecimalLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimeLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimestampLogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
+import org.joda.time.DateTimeZone;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static io.trino.parquet.ParquetTypeUtils.createDecimalType;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static java.util.Objects.requireNonNull;
+
+public final class ColumnReaderFactory
+{
+    private ColumnReaderFactory() {}
+
+    public static ColumnReader create(PrimitiveField field, DateTimeZone timeZone)
+    {
+        PrimitiveTypeName primitiveTypeName = field.getDescriptor().getPrimitiveType().getPrimitiveTypeName();
+        List<ColumnReaderProvider> providers = getColumnReaderProviders(primitiveTypeName);
+        requireNonNull(providers, "Unsupported Parquet column type: " + primitiveTypeName);
+
+        return providers.stream()
+                .filter(provider -> provider.accepts(field))
+                .findFirst()
+                .map(provider -> provider.create(field, timeZone))
+                .orElseThrow(() -> new TrinoException(NOT_SUPPORTED, "Unsupported Parquet column type: " + primitiveTypeName));
+    }
+
+    // Providers will be iterated first to last in order to find a matching one.
+    // Lists should be sorted from the most specific to the most generic reader.
+    private static List<ColumnReaderProvider> getColumnReaderProviders(PrimitiveTypeName primitiveTypeName)
+    {
+        return switch (primitiveTypeName) {
```

Comment:
I've updated this to use constant ColumnReaderProviders instead of creating them each time.
I would prefer to keep the switch case instead of Map though because it provides a compile time check that all cases of `primitiveTypeName` are covered.

Reply:
Ok, that's reasonable.

Reply:
However, something I've been thinking about is that we should be resolving the providers by first looking at the Trino type and then, based on the parquet type w/ annotations decide which column reader is most appropriate (or fail if none are, as that would indicate there's an incompatibility between the declared trino type and the stored data).

The current approach relies on the column readers behaving "properly" when lumped together with the column readers for other types -- i.e., that they won't inadvertently claim they can decode a type they are not supposed to. It also requires "sorting from the most specific to the most generic reader" as described in the comment above, but that may not always be possible.

Reply:
I'm not sure that checking Trino type first would simplify things. E.g. Trino INTEGER type block could be produced by a short decimal of scale 0 stored in INT32, INT64, BINARY or FIXED_LEN_BYTE_ARRAY.
The sorting of column providers is really only necessary where we have a case of multiple column providers accepting the same parquet and trino type without applying specific annotation checks.
After separation of the batched and non-batched switch cases, there are a few cases (Time and zero scale decimal for INT32 and zero scale decimal for INT64) where the sorting is no longer sufficient. I've now added checks on specific type annotation (or lack of annotation) in `accepts` method of the column providers to further tighten up what the column providers will match.

Reply:
On the other hand, doing the lookup first by Trino type would make it clear exactly how a given type is mapped. Currently, that information is scattered across multiple places, and figuring it out requires understanding the order in which the providers are checked, the specific logic inside each of the accepts methods, etc. The pattern of letting providers decide whether they match is overkill for this case -- there's no ambiguity once the trino type and the parquet type with annotations are known. That approach would be suitable if there really were a set of preferences or priorities (think content negotiation in HTTP, for instance)

Reply:
To illustrate, the mapping and creation of readers could be encapsulated in a single, much simpler method that does something along the lines of:

```java
    public static ColumnReader create(PrimitiveField field, DateTimeZone timeZone)
    {
        Type type = field.getType();
        PrimitiveTypeName primitiveType = field.getDescriptor().getPrimitiveType().getPrimitiveTypeName();
        LogicalTypeAnnotation annotation = field.getDescriptor().getPrimitiveType().getLogicalTypeAnnotation();

        if (type instanceof BooleanType && primitiveType == PrimitiveTypeName.BOOLEAN) {
            return new BooleanColumnReader(field);
        }
        else if (type instanceof IntegerType && primitiveType == PrimitiveTypeName.INT32) {
            return new IntColumnReader(field);
        }
        else if (type instanceof BigintType && primitiveType == PrimitiveTypeName.INT64) {
            return new LongColumnReader(field);
        }
        else if (type instanceof RealType && primitiveType == PrimitiveTypeName.FLOAT) {
            return new FloatColumnReader(field);
        }
        else if (type instanceof DoubleType && primitiveType == PrimitiveTypeName.DOUBLE) {
            return new DoubleColumnReader(field);
        }
        else if (type instanceof UuidType && primitiveType == PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY) {
            return new UuidColumnReader(field);
        }
        else if (type instanceof VarcharType && primitiveType == PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY) {
            return new BinaryColumnReader(field);
        }
        else if (type instanceof VarbinaryType && primitiveType == PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY) {
            return new BinaryColumnReader(field);
        }
        else if (type instanceof CharType && primitiveType == PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY) {
            return new BinaryColumnReader(field);
        }
        else if (type instanceof TimestampType timestampType) {
            if (primitiveType == PrimitiveTypeName.INT64) {
                if (annotation instanceof TimestampLogicalTypeAnnotation timestampAnnotation) {
                    return switch (timestampAnnotation.getUnit()) {
                        case MILLIS -> new Int64TimestampMillisColumnReader(field);
                        case MICROS -> new TimestampMicrosColumnReader(field);
                        case NANOS -> new Int64TimestampNanosColumnReader(field);
                    };
                }
            }
            else if (primitiveType == PrimitiveTypeName.INT96) {
                return new TimestampColumnReader(field, timeZone);
            }
        }
        else if (type instanceof TimeType) {
            if (annotation instanceof TimeLogicalTypeAnnotation timeAnnotation && timeAnnotation.getUnit() == LogicalTypeAnnotation.TimeUnit.MICROS) {
                return new TimeMicrosColumnReader(field);
            }
        }
        else if (type instanceof DecimalType && annotation instanceof DecimalLogicalTypeAnnotation) {
            return DecimalColumnReaderFactory.createReader(field, (DecimalType) type);
        }

        throw new TrinoException(NOT_SUPPORTED, "...");
    }
```

Reply:
I've rewritten this part now. I've removed `ColumnReaderProvider` entirely. For the batched readers scenario, I've used trino type first to find the column reader.
For the fallback code path, I've kept it the same as pre-existing code so that there is no risk of regression if we disable the optimized reads code path.